### PR TITLE
Need a short description (OOPS!).

### PR DESCRIPTION
**Pull request by chgibb-apple**
#### 16a15e2b4283dbea29f3a4cf9b6f4fd41b34c72e
<pre>
Need a short description (OOPS!).
Need the bug URL (OOPS!).

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/delete-if-too-large:
</pre>

### CODE REVIEW

**Comment by JonWBedard** [comment]
> #### [ed091f5](https://github.com/WebKit/WebKit/commit/ed091f5dd65e4a646ac17cd82b7237d1b0319657)
> Prune coresymbolicationd cache if it grows larger than 10G

We do want a formatted commit message, https://github.com/WebKit/WebKit/blob/main/Introduction.md#commit-messages

File: Tools/CISupport/Shared/maybe-delete-directory.bash [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
```
Comment (aproskuryakov) [inline Tools/CISupport/Shared/maybe-delete-directory.bash]: Nit: we'd normally use `.sh` extension for the file. the shebang looks fine.

Slightly more substantial, please use `/bin/sh`. The macOS version of `bash` is very old, and actually prints out warnings about pending removal if I remember correctly.

We'd probably use Python for something like this, but I think that shell is OK.

File: Tools/CISupport/Shared/maybe-delete-directory.bash [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+# Deletes the directory given as the first argument if and only if it is greater than 10GB in size
+
+if [[ $# < 1 ]]
+then
+    echo "Must provide absolute path to directory to prune."
```
Comment (aproskuryakov) [inline Tools/CISupport/Shared/maybe-delete-directory.bash]: Nit: it would be nicer to print this out as "Usage" help,  as opposed to having the computer give orders to humans :)

File: Tools/CISupport/Shared/maybe-delete-directory.bash [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    exit 1
+fi
+
+if [[ -L "${1%/}" ]]
+then
+    echo "$1 must not be a symbolic link."
```
Comment (aproskuryakov) [inline Tools/CISupport/Shared/maybe-delete-directory.bash]: Similarly, I'd have the script say that it `$1` is a symlink, please provide a direct directory link (not quite sure why symlinks aren't OK though).

File: Tools/CISupport/build-webkit-org/factories.py [hunk 4]
```diff
@@ -38,6 +38,7 @@ def __init__(self, platform, configuration, architectures, buildOnly, additional
             self.addStep(KillOldProcesses())
         self.addStep(CleanBuildIfScheduled())
         self.addStep(DeleteStaleBuildFiles())
+        self.addStep(MaybePruneCoreSymbolicationdCache())
```
Comment (aproskuryakov) [inline Tools/CISupport/build-webkit-org/factories.py]: Not sure if "Maybe" adds valuable information to the step name. How is it more conditional than existing steps?

File: Tools/CISupport/build-webkit-org/steps.py [hunk 5]
```diff
@@ -203,6 +203,13 @@ class KillOldProcesses(shell.Compile):
     description = ["killing old processes"]
     descriptionDone = ["killed old processes"]
     command = ["python3", "Tools/CISupport/kill-old-processes", "buildbot"]
+    
+
+class MaybePruneCoreSymbolicationdCache(shell.Compile):
+    name = "maybe-prune-coresymbolicationd-cache"
+    description = ["pruning coresymbolicationd cache if it is over 10GB in size"]
```
Comment (aproskuryakov) [inline Tools/CISupport/build-webkit-org/steps.py]: "pruning coresymbolicationd cache to < 10GB" is shorter wording for your consideration.

**Review by aproskuryakov — COMMENTED** [review]
Added some minor comments. Probably best for someone other than me to do final review.

File: Tools/CISupport/build-webkit-org/factories.py (hunk 4 above)
Comment (aproskuryakov) [reply Tools/CISupport/build-webkit-org/factories.py]: Also, I think that this script would be executed on non-Mac queues too (Windows and Linux). We probably want to make it conditional on Mac being the host platform (keeping iOS Simulator in mind).

File: Tools/CISupport/build-webkit-org/steps.py [hunk 6]
```diff
@@ -203,6 +203,13 @@ class KillOldProcesses(shell.Compile):
...
+    descriptionDone = ["pruned coresymbolicationd cache"]
+    command = ["bash", "Tools/CISupport/Shared/maybe-delete-directory.bash", "/System/Library/Caches/com.apple.coresymbolicationd"]
```
Comment (aproskuryakov) [inline Tools/CISupport/build-webkit-org/steps.py]: With the shebang and executable bit set, this can be just `command = ["Tools/CISupport/Shared/maybe-delete-directory.bash", ...`.

File: Tools/CISupport/Shared/maybe-delete-directory.bash [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+fi
+
+echo "Considering directory $1 for deletion"
+
+IFS=$'\n' read -r -d '' sizeInBytesGreaterThanThreshold < <( \
+ du -xPc $1 | \
+ awk ' { last = $0; } END { print last; } ' | \
+ awk ' { print $1; }' | \
+ awk ' { if ( $1 > 10737418240) { print "true"; } else { print "false"; } } ' && \
+ printf '\0')
```
Comment (JonWBedard) [inline Tools/CISupport/Shared/maybe-delete-directory.bash]: I don't think I'm going to be able to follow through on reviewing this, so this should be taken as a possible suggestion rather than request.

One of the things @aj062 and I have done with some more complicated buildbot steps is use `||` in a shell command to conditionally run a single command in a buildbot step without a stand-alone script. It results in the active code being quite ugly, but it's also nice because it lets us have a step that's detached from the state of the checkout and works regardless of the branch we're running on. This particular bit of code is guaranteed to have a checkout, so that isn't strictly necessary, but also, we have single check followed by a single command that is only relevant on Mac, which fits the shell-command from buildbot pattern.

File: Tools/CISupport/build-webkit-org/factories.py (hunk 4 above)
Comment (chgibb-apple) [reply Tools/CISupport/build-webkit-org/factories.py]: A "Maybe" prefix is a naming pattern I've used in the past to describe an operation that may (or may not) cause an observable side effect. I've renamed this class to make it more clear.

I've also added a condition to this step to only run on Darwin hosts.

File: Tools/CISupport/Shared/maybe-delete-directory.bash (hunk 1 above)
Comment (chgibb-apple) [reply Tools/CISupport/Shared/maybe-delete-directory.bash]: `bash` on macOS prints a deprecation warning when invoked interactively but does not when invoked non-interactively (such as by `bash foo.bash`). This scripts uses a number of bash-isms that do not work in `sh`. This script does appear to work in `zsh` however. I've changed the shebang and file extension to `zsh`.

Comment (aproskuryakov) [reply Tools/CISupport/Shared/maybe-delete-directory.bash]: No, please don't use zsh either. We standardized on /bin/sh shebang intentionally.

Comment (chgibb-apple) [reply Tools/CISupport/Shared/maybe-delete-directory.bash]: Okay. I will rewrite this script to remove the bash-isms that `sh` does not support.

File: Tools/CISupport/build-webkit-org/steps.py [hunk 8]
```diff
@@ -205,6 +205,13 @@ class KillOldProcesses(shell.Compile):
     command = ["python3", "Tools/CISupport/kill-old-processes", "buildbot"]
 
 
+class PruneCoreSymbolicationdCacheIfTooLarge(shell.Compile):
+    name = "prune-coresymbolicationd-cache-if-too-large"
+    description = ["pruning coresymbolicationd cache to < 10GB"]
+    descriptionDone = ["pruned coresymbolicationd cache"]
+    command = ["zsh", "Tools/CISupport/Shared/maybe-delete-directory.zsh", "/System/Library/Caches/com.apple.coresymbolicationd"]
```
Comment (roynos) [inline Tools/CISupport/build-webkit-org/steps.py]: where is `maybe-delete-directory.zsh`? Is it a separate patch?

Comment (roynos) [reply Tools/CISupport/build-webkit-org/steps.py]: I don't know what the script does. I would delete the sub directories. Instead of deleting the folder itself.

File: Tools/CISupport/Shared/maybe-delete-directory.bash (hunk 7 above)
Comment (chgibb-apple) [reply Tools/CISupport/Shared/maybe-delete-directory.bash]: Can you point me towards some prior art demonstrating what you've described please? I'm not sure I understand the pattern you've described.

Comment (JonWBedard) [reply Tools/CISupport/Shared/maybe-delete-directory.bash]: In `Tools/CISupport/ews-build/steps.py`, if you search `self.shell_command`, you can see the places where we're running ugly compound commands. The reason I call this possibility out is that the code as proposed requires us to cherry-pick this change to a branch, probably two branches, actually. Which is certainly a thing we can do, it just increases the cost of a separate file in a buildbot step like this.

File: Tools/CISupport/Shared/maybe-delete-directory.sh [hunk 9]
```diff
@@ -0,0 +1,48 @@
+#!/bin/sh
```
Comment (roynos) [inline Tools/CISupport/Shared/maybe-delete-directory.sh]: most of our multiline scripts are written in Python. I am not opposed to landing this script as a bash script.

File: Tools/CISupport/build-webkit-org/factories.py (hunk 4 above)
Comment (roynos) [reply Tools/CISupport/build-webkit-org/factories.py]: We've used "IfNecessary" suffix.

File: Tools/CISupport/build-webkit-org/steps.py (hunk 8 above)
Comment (roynos) [reply Tools/CISupport/build-webkit-org/steps.py]: have you considered using `find` with `atime` to target the directories that no one is accessing. this should prevent applications from crashing because files and folders no longer exist.

File: Tools/CISupport/Shared/maybe-delete-directory.bash (hunk 7 above)
Comment (chgibb-apple) [reply Tools/CISupport/Shared/maybe-delete-directory.bash]: Using [`CheckOutSpecificRevision`](https://github.com/WebKit/WebKit/blob/main/Tools/CISupport/ews-build/steps.py#L678) and the buildbot documentation as an example, I think that structure mostly makes sense. Is there a clean way to run a specific step locally? Assuming I would like to rewrite `PruneCoreSymbolicationdCacheIfTooLarge` as a `shell.Command` and run just that locally, how would I go about doing that?

With respect to unit testing, is there a preferred way to unit test `shell.Command` subclasses?

Comment (JonWBedard) [reply Tools/CISupport/Shared/maybe-delete-directory.bash]: I think you want to try and break down that script into 2 commands and join them with `&&`, the first being the check and the second being the removal. I would make sure that the combinations run locally in your terminal. Next step is running a test on our staging instance, ews-build.webkit-uat.org. That's a real buildbot instance which a few machines attached. You want to have it do a build that runs your change (cc @aj062 to help). Last step would be creating a unit test, which is mostly just going to be a string check. `CheckOutSpecificRevision` has unit tests in in `steps_unittest.py`, my suggestion is you copy and mutate those because that step has relatively simple tests.

Comment (chgibb-apple) [reply Tools/CISupport/Shared/maybe-delete-directory.bash]: I've rewritten all the logic in Python instead of stretching to use shell. I tried to use `getProperty`/`setProperty` to inject fakes for `Path` and some other functionality to try to make this testable. It appears this does not work because properties have to be JSON serialisable.

File: Tools/CISupport/build-webkit-org/steps.py [hunk 10]
```diff
@@ -205,6 +206,38 @@ class KillOldProcesses(shell.Compile):
...
+    descriptionDone = ["pruned coresymbolicationd cache"]
+    flunkOnFailure = True
+    haltOnFailure = True
+
+    def __init__(self):
+        buildstep.BuildStep.__init__(self)
+
+    def start(self):
+        directory = '/System/Library/Caches/com.apple.coresymbolicationd'
+        directoryPath = Path(directory).resolve()
```
Comment (JonWBedard) [inline Tools/CISupport/build-webkit-org/steps.py]: Does this actually operate on the worker? cc @aj062 to double check that If so, I _love_ the approach!

File: Tools/CISupport/build-webkit-org/steps.py [hunk 11]
```diff
@@ -205,6 +206,38 @@ class KillOldProcesses(shell.Compile):
     command = ["python3", "Tools/CISupport/kill-old-processes", "buildbot"]
 
 
+class PruneCoreSymbolicationdCacheIfTooLarge(buildstep.BuildStep):
```
Comment (JonWBedard) [inline Tools/CISupport/build-webkit-org/steps.py]: ~Generally, this is only relevant for Mac queues, right?~

Comment (JonWBedard) [reply Tools/CISupport/build-webkit-org/steps.py]: ~Might want to add a `doStepIf`~

Comment (JonWBedard) [reply Tools/CISupport/build-webkit-org/steps.py]: Looks like you did that in the factory

File: Tools/CISupport/build-webkit-org/steps.py (hunk 10 above)
Comment (aproskuryakov) [reply Tools/CISupport/build-webkit-org/steps.py]: Indeed, let's confirm that! I suspect that it's probably running on the master.

File: Tools/CISupport/build-webkit-org/steps.py [hunk 12]
```diff
@@ -205,6 +206,38 @@ class KillOldProcesses(shell.Compile):
...
+    name = "prune-coresymbolicationd-cache-if-too-large"
+    description = ["pruning coresymbolicationd cache to < 10GB"]
+    descriptionDone = ["pruned coresymbolicationd cache"]
+    flunkOnFailure = True
```
Comment (aj062) [inline Tools/CISupport/build-webkit-org/steps.py]: Do we really want to fail the build (and have red ews status-bubbles) if this step fails? 
Seems like this and haltOnFailure should be no.

File: Tools/CISupport/build-webkit-org/steps.py (hunk 10 above)
Comment (aj062) [reply Tools/CISupport/build-webkit-org/steps.py]: Yes, I think it would run on master.
For example this email code (`send_email_for_git_issue`) in `start()` also run on server: https://github.com/WebKit/WebKit/blob/main/Tools/CISupport/ews-build/steps.py#L658

File: Tools/CISupport/build-webkit-org/steps.py [hunk 13]
```diff
@@ -205,6 +206,38 @@ class KillOldProcesses(shell.Compile):
...
+            self._addToLog('stdio', 'coresymbolicationd cache does not exist')
+            self.finished(SKIPPED)
+            return None
+
+        if not directoryPath.is_dir():
+            self._addToLog('stdio', 'coresymbolicationd cache is not a directory')
+            self.finished(FAILURE)
+            return None
+
+        if directoryPath.stat().st_size >= 10737418240:
```
Comment (aj062) [inline Tools/CISupport/build-webkit-org/steps.py]: Better to make this number a variable with readable name, and also have it explicitly as a multiplication to indicate what it represent.

File: Tools/CISupport/build-webkit-org/steps.py (hunk 12 above)
Comment (chgibb-apple) [reply Tools/CISupport/build-webkit-org/steps.py]: Both have been changed to `False`.

File: Tools/CISupport/build-webkit-org/steps.py (hunk 13 above)
Comment (chgibb-apple) [reply Tools/CISupport/build-webkit-org/steps.py]: Done.

**Review by aj062 — CHANGES_REQUESTED** [review]
r- since this would run the commands on master and not on the bots. It might not be straight-forward to do the directory size checks on worker and delete the directory on worker in ShellCommands. I'm fine with the previous script approach (adding a shell/python script which would run on worker) as that would be more flexible and easily extendable in future. 

Can also make this script generic to take care of any kind of deletion (and on multiple platforms), for example it can even be extended in future to include the fix for <rdar://problem/72849252>.

File: Tools/CISupport/build-webkit-org/steps.py [hunk 14]
```diff
@@ -205,6 +206,39 @@ class KillOldProcesses(shell.Compile):
...
+            self._addToLog('stdio', 'coresymbolicationd cache does not exist')
+            self.finished(SKIPPED)
+            return None
+
+        if not directoryPath.is_dir():
+            self._addToLog('stdio', 'coresymbolicationd cache is not a directory')
+            self.finished(FAILURE)
+            return None
+
+        tenGigabytes = 1024 * 1000 * 1000 * 10
```
Comment (aj062) [inline Tools/CISupport/build-webkit-org/steps.py]: Nit: I guess you meant 1024 instead of 1000

Comment (chgibb-apple) [reply Tools/CISupport/build-webkit-org/steps.py]: Fixed.

File: Tools/Scripts/delete-if-too-large [hunk 15]
```diff
@@ -0,0 +1,25 @@
+import sys
```
Comment (aj062) [inline Tools/Scripts/delete-if-too-large:26]: shebang missing

File: Tools/Scripts/delete-if-too-large [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+        return 1
+
+    if not directoryPath.is_dir():
+        print(directory + "is not a directory")
+        return 1
+
+    tenGigabytes = 1024 * 1024 * 1024 * 10
+    if directoryPath.stat().st_size >= tenGigabytes:
+        os.remove(directory)
+        print("Successfully removed " + directory)
```
Comment (aj062) [inline Tools/Scripts/delete-if-too-large]: Maybe better to say: f"Deleted {directory}" (instead of using the word successfully)

Comment (chgibb-apple) [reply Tools/Scripts/delete-if-too-large]: Done.

File: Tools/Scripts/delete-if-too-large (hunk 15 above)
Comment (chgibb-apple) [reply Tools/Scripts/delete-if-too-large:26]: Fixed.

File: Tools/Scripts/delete-if-too-large [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+import os
+from pathlib import Path
+
+
+def main():
+    directory = sys.argv[1]
+    directoryPath = Path(directory).resolve()
+
+    if not directoryPath.exists():
+        print(directory + "does not exist")
```
Comment (roynos) [inline Tools/Scripts/delete-if-too-large:36]: this will `directory_namedoes not exist` I recommend using `.format()` or `f strings`. since this is python3 specific you should use strings.

File: Tools/Scripts/delete-if-too-large [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+        return 1
+
+    if not directoryPath.is_dir():
+        print(directory + "is not a directory")
+        return 1
```
Comment (roynos) [inline Tools/Scripts/delete-if-too-large:41]: nit: I would change the return value here. you can check against the return value without parsing standard output.

File: Tools/Scripts/delete-if-too-large [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+
+    tenGigabytes = 1024 * 1024 * 1024 * 10
+    if directoryPath.stat().st_size >= tenGigabytes:
+        os.remove(directory)
+        print("Deleted " + directory)
+
```
Comment (roynos) [inline Tools/Scripts/delete-if-too-large:47]: `return 0` not sure what sys.exit(None) resolves as.

File: Tools/Scripts/delete-if-too-large [hunk 20]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python3
+
+import sys
+import os
+from pathlib import Path
+
+
+def main():
+    directory = sys.argv[1]
```
Comment (roynos) [inline Tools/Scripts/delete-if-too-large:32]: use `argparse` ?

File: Tools/Scripts/delete-if-too-large [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    if not directoryPath.exists():
+        print(directory + "does not exist")
+        return 1
+
+    if not directoryPath.is_dir():
+        print(directory + "is not a directory")
+        return 1
+
+    tenGigabytes = 1024 * 1024 * 1024 * 10
+    if directoryPath.stat().st_size >= tenGigabytes:
```
Comment (roynos) [inline Tools/Scripts/delete-if-too-large:44]: I would make this argument and call it something like `size_threshold`. I am only saying this because the name is generic `delete-if-too-large`

File: Tools/Scripts/delete-if-too-large [hunk 22]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python3
```
Comment (roynos) [inline Tools/Scripts/delete-if-too-large:1]: not sure how to check this via GitHub. but did you make this script executable?

Comment (aj062) [inline Tools/Scripts/delete-if-too-large:1]: Please add Copyright message (you can copy it from another similar file).

**Review by aj062 — APPROVED** [review]

File: Tools/CISupport/Shared/maybe-delete-directory.sh (hunk 9 above)
Comment (chgibb-apple) [reply Tools/CISupport/Shared/maybe-delete-directory.sh]: This was rewritten as Python.

File: Tools/CISupport/build-webkit-org/steps.py (hunk 8 above)
Comment (chgibb-apple) [reply Tools/CISupport/build-webkit-org/steps.py]: This was rewritten as Python.

File: Tools/Scripts/delete-if-too-large (hunk 17 above)
Comment (chgibb-apple) [reply Tools/Scripts/delete-if-too-large:36]: Fixed. Thank you!

File: Tools/Scripts/delete-if-too-large (hunk 19 above)
Comment (chgibb-apple) [reply Tools/Scripts/delete-if-too-large:47]: Fixed. Thank you!

File: Tools/Scripts/delete-if-too-large (hunk 20 above)
Comment (chgibb-apple) [reply Tools/Scripts/delete-if-too-large:32]: What does `argparse` provide over using `sys.argv`?

Comment (roynos) [reply Tools/Scripts/delete-if-too-large:32]: it will make it easier to add arguments, description, and help output.

File: Tools/Scripts/delete-if-too-large (hunk 19 above)
Comment (roynos) [reply Tools/Scripts/delete-if-too-large:47]: Fixed in a new PR?
